Sometimes it is vital to buy pills as soon as possible. Following a conducted survey, the average population served by a pharmacy is 4 182 inhabitants per one. The number indicates a country`s community infrastructure and capacity, as well as the accessibility of pharmacy services and, in total access to medicines. Some towns have a low density of the population, while a large population. The question where a new pharmacy may be opened is also a part of its business plan. As a result, we can see, that questions, is it enough quantity of pharmacies, and where a new should be opened is of high interest.
At the beginning I will discover census data of Edmonton. The data is imported from official source https://data.edmonton.ca/

In [1]:
import numpy as np # library to handle data in a vectorized manner
import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json # library to handle JSON files

In [2]:
# import 2016 Census data for Edmonton neighborhoods - Population by Age Range 
df_nb = pd.read_csv('https://data.edmonton.ca/api/views/phd4-y42v/rows.csv?accessType=DOWNLOAD')
df_nb.head ()

Ward  Neighbourhood Number Neighbourhood Name  0 - 4  5 - 9  10 - 14  \
0   WARD 1                  3140          CRESTWOOD     68     83       97   
1   WARD 1                  3330           PARKVIEW    126    162      138   
2  WARD 11                  6110         CPR IRVINE      0      1        0   
3   WARD 9                  5350     RHATIGAN RIDGE     73    130      117   
4   WARD 5                  4140            ELMWOOD    131    121      130   

   15 - 19  20 - 24  25 - 29  30 - 34  35 - 39  40 - 44  45 - 49  50 - 54  \
0       99       76       62       59       74       77      110      116   
1      165      155       96      128      149      164      164      218   
2        0       18       44       39        4        1        1        6   
3      112      112       84       56       83      107      103      142   
4      169      180      185      173      135      154      154      161   

   55 - 59  60 - 64  65 - 69  70 - 74  75 - 79  80 - 84  85+  No Response  
0      148       98      100       91       80       64   49          800  
1      247      197      146       89       90       82   64          705  
2        1        0        1        0        0        2    0           48  
3      185      205      171      102       74       42   39         1140  
4      133      145      140       92      117       60   51          290

The imported data includes the numbers and the names of Neighbourhoods, as well as quantity of population devided by age groups. Let`s clean the data.

In [3]:
# Remove excessive columns 'Ward' and 'Neighbourhood Number'
df_nb = df_nb.drop(['Ward', 'Neighbourhood Name'], axis=1)
df_nb.head ()

Neighbourhood Number  0 - 4  5 - 9  10 - 14  15 - 19  20 - 24  25 - 29  \
0                  3140     68     83       97       99       76       62   
1                  3330    126    162      138      165      155       96   
2                  6110      0      1        0        0       18       44   
3                  5350     73    130      117      112      112       84   
4                  4140    131    121      130      169      180      185   

   30 - 34  35 - 39  40 - 44  45 - 49  50 - 54  55 - 59  60 - 64  65 - 69  \
0       59       74       77      110      116      148       98      100   
1      128      149      164      164      218      247      197      146   
2       39        4        1        1        6        1        0        1   
3       56       83      107      103      142      185      205      171   
4      173      135      154      154      161      133      145      140   

   70 - 74  75 - 79  80 - 84  85+  No Response  
0       91       80       64   49          800  
1       89       90       82   64          705  
2        0        0        2    0           48  
3      102       74       42   39         1140  
4       92      117       60   51          290

As we can see from the table there are no any information about the geographical position of the Neighbourhoods. This information can be found at the same source from the other table. Lets import the table in the notebook.

In [4]:
# Getting information on Edmonton neighborhoods Latitude and Longitude
Edm_nb = pd.read_csv('https://data.edmonton.ca/api/views/3b6m-fezs/rows.csv?accessType=DOWNLOAD')
Edm_nb.head ()

Neighbourhood Number        Neighbourhood Name  Area Sq Km   Latitude  \
0                  6790                 Wild Rose    1.772290  53.470564   
1                  1200     River Valley Victoria    1.298977  53.534988   
2                  4330       Mitchell Industrial    0.592507  53.574016   
3                  5350            Rhatigan Ridge    1.344078  53.474506   
4                  1280  Yellowhead Corridor West    1.164872  53.582841   

    Longitude                                   Location  \
0 -113.381167    (53.47056409015751, -113.3811672388747)   
1 -113.523686   (53.53498830876779, -113.52368603311965)   
2 -113.584201  (53.574015777213816, -113.58420113302314)   
3 -113.587569   (53.474506145469945, -113.5875691214093)   
4 -113.512729   (53.58284074814297, -113.51272869912015)   

   Neighbourhood Boundaries : 2019  Roadway Maintenance Area Polygon  \
0                               74                             135.0   
1                              344                             144.0   
2                              112                             206.0   
3                               71                              64.0   
4                              257                             223.0   

   Edmonton Public School Board (EPSB) Ward Boundaries (effective at 12:00 AM on Oct 16, 2017)  \
0                                                7.0                                             
1                                                4.0                                             
2                                                3.0                                             
3                                                6.0                                             
4                                                3.0                                             

   Edmonton Catholic School District Ward Boundaries (effective at 12:00 AM on Oct 16, 2017)  \
0                                                7.0                                           
1                                                5.0                                           
2                                                1.0                                           
3                                                6.0                                           
4                                                1.0                                           

   City of Edmonton - Ward Boundaries (effective at 12:00 AM on Oct 16, 2017)  \
0                                                4.0                            
1                                                9.0                            
2                                                5.0                            
3                                               12.0                            
4                                                5.0                            

   :@computed_region_eq8d_jmrp  :@computed_region_da6r_6gkw  \
0                        381.0                         12.0   
1                        308.0                          6.0   
2                        251.0                          2.0   
3                        284.0                          9.0   
4                        392.0                          2.0   

   :@computed_region_3z8d_fhyb  
0                          NaN  
1                          NaN  
2                          NaN  
3                          NaN  
4                          NaN

Now I merge this two Dataframes into one and clean it from the excessive columns, that I do not need in this research (Neighbourhood Boundaries, Roadway Maintenance Area Polygon, etc.).

In [5]:
# now merge two dataframes and cut the excessive columns
result_Ed = pd.merge(left=df_nb, right=Edm_nb, how='left', left_on='Neighbourhood Number', right_on='Neighbourhood Number')
result_Ed = result_Ed.drop ([':@computed_region_eq8d_jmrp', 
                             ':@computed_region_da6r_6gkw', 
                             ':@computed_region_3z8d_fhyb', 
                             'City of Edmonton - Ward Boundaries (effective at 12:00 AM on Oct 16, 2017)',
                            'Edmonton Catholic School District Ward Boundaries (effective at 12:00 AM on Oct 16, 2017)',
                            'Edmonton Public School Board (EPSB) Ward Boundaries (effective at 12:00 AM on Oct 16, 2017)',
                            'Roadway Maintenance Area Polygon',
                            'Neighbourhood Boundaries : 2019',
                            'Location'], axis = 1)
result_Ed.head ()

Neighbourhood Number  0 - 4  5 - 9  10 - 14  15 - 19  20 - 24  25 - 29  \
0                  3140     68     83       97       99       76       62   
1                  3330    126    162      138      165      155       96   
2                  6110      0      1        0        0       18       44   
3                  5350     73    130      117      112      112       84   
4                  4140    131    121      130      169      180      185   

   30 - 34  35 - 39  40 - 44  45 - 49  50 - 54  55 - 59  60 - 64  65 - 69  \
0       59       74       77      110      116      148       98      100   
1      128      149      164      164      218      247      197      146   
2       39        4        1        1        6        1        0        1   
3       56       83      107      103      142      185      205      171   
4      173      135      154      154      161      133      145      140   

   70 - 74  75 - 79  80 - 84  85+  No Response Neighbourhood Name  Area Sq Km  \
0       91       80       64   49          800          Crestwood    1.168158   
1       89       90       82   64          705           Parkview    1.546448   
2        0        0        2    0           48         CPR Irvine    0.663610   
3      102       74       42   39         1140     Rhatigan Ridge    1.344078   
4       92      117       60   51          290            Elmwood    1.025925   

    Latitude   Longitude  
0  53.535434 -113.569038  
1  53.524060 -113.567914  
2  53.507527 -113.490549  
3  53.474506 -113.587569  
4  53.515738 -113.605993

I recieved the table that contains all the data about Neighbourhoods in Edmonton. It includes Neighbourhood`s names, inhabitants, coordinates, area. 
Some cells do not contain coordinates. I will cut them out of the table because they can not be used in my research.

In [6]:
# clean rows without information about Latitude and Longitude
result_a = result_Ed.dropna(subset=['Latitude', 'Longitude'])
result_a.head ()

Neighbourhood Number  0 - 4  5 - 9  10 - 14  15 - 19  20 - 24  25 - 29  \
0                  3140     68     83       97       99       76       62   
1                  3330    126    162      138      165      155       96   
2                  6110      0      1        0        0       18       44   
3                  5350     73    130      117      112      112       84   
4                  4140    131    121      130      169      180      185   

   30 - 34  35 - 39  40 - 44  45 - 49  50 - 54  55 - 59  60 - 64  65 - 69  \
0       59       74       77      110      116      148       98      100   
1      128      149      164      164      218      247      197      146   
2       39        4        1        1        6        1        0        1   
3       56       83      107      103      142      185      205      171   
4      173      135      154      154      161      133      145      140   

   70 - 74  75 - 79  80 - 84  85+  No Response Neighbourhood Name  Area Sq Km  \
0       91       80       64   49          800          Crestwood    1.168158   
1       89       90       82   64          705           Parkview    1.546448   
2        0        0        2    0           48         CPR Irvine    0.663610   
3      102       74       42   39         1140     Rhatigan Ridge    1.344078   
4       92      117       60   51          290            Elmwood    1.025925   

    Latitude   Longitude  
0  53.535434 -113.569038  
1  53.524060 -113.567914  
2  53.507527 -113.490549  
3  53.474506 -113.587569  
4  53.515738 -113.605993

The information about inhabitants age groups could be used as variable in determining of how many consumers an pharmacy could have. Infortinately, I did not find in the Web a correlation between age and money a person spend for drugs. Subsequently, I decided to calculate sum of inhabitants of the Neighbourhoods in Edmonton.

In [7]:
# get the sum of inhabitants
result_a['Inhabitants'] = result_a.iloc[:, 1:20].sum(1)
result_a.head ()

C:\Users\avh77\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


Neighbourhood Number  0 - 4  5 - 9  10 - 14  15 - 19  20 - 24  25 - 29  \
0                  3140     68     83       97       99       76       62   
1                  3330    126    162      138      165      155       96   
2                  6110      0      1        0        0       18       44   
3                  5350     73    130      117      112      112       84   
4                  4140    131    121      130      169      180      185   

   30 - 34  35 - 39  40 - 44  45 - 49  50 - 54  55 - 59  60 - 64  65 - 69  \
0       59       74       77      110      116      148       98      100   
1      128      149      164      164      218      247      197      146   
2       39        4        1        1        6        1        0        1   
3       56       83      107      103      142      185      205      171   
4      173      135      154      154      161      133      145      140   

   70 - 74  75 - 79  80 - 84  85+  No Response Neighbourhood Name  Area Sq Km  \
0       91       80       64   49          800          Crestwood    1.168158   
1       89       90       82   64          705           Parkview    1.546448   
2        0        0        2    0           48         CPR Irvine    0.663610   
3      102       74       42   39         1140     Rhatigan Ridge    1.344078   
4       92      117       60   51          290            Elmwood    1.025925   

    Latitude   Longitude  Inhabitants  
0  53.535434 -113.569038         2351  
1  53.524060 -113.567914         3285  
2  53.507527 -113.490549          166  
3  53.474506 -113.587569         3077  
4  53.515738 -113.605993         2721

In [8]:
# drop the excessive columns and rows without inhabitants
result_a = result_a.drop (['0 - 4', 
                           '5 - 9', 
                           '10 - 14', 
                           '15 - 19', 
                           '20 - 24', 
                           '25 - 29',
                           '30 - 34',
                           '35 - 39',
                           '40 - 44',
                           '45 - 49',
                           '50 - 54',
                           '55 - 59',
                           '60 - 64',
                           '65 - 69',
                           '70 - 74',
                           '75 - 79',
                           '80 - 84',
                           '85+',
                           'No Response'], axis = 1)
result_a = result_a[result_a['Inhabitants'] != 0]
result_a.head ()

Neighbourhood Number Neighbourhood Name  Area Sq Km   Latitude   Longitude  \
0                  3140          Crestwood    1.168158  53.535434 -113.569038   
1                  3330           Parkview    1.546448  53.524060 -113.567914   
2                  6110         CPR Irvine    0.663610  53.507527 -113.490549   
3                  5350     Rhatigan Ridge    1.344078  53.474506 -113.587569   
4                  4140            Elmwood    1.025925  53.515738 -113.605993   

   Inhabitants  
0         2351  
1         3285  
2          166  
3         3077  
4         2721

I recieved the DataFrame that contains all necessary information for estimation of distribution of inhabitants over the Edmonton area. Lets visualize the information on the map of the town.   

In [10]:
!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
# import k-means from clustering stage
from sklearn.cluster import KMeans
!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\avh77\Anaconda3

  added / updated specs:
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2019.3.9           |           py37_0         149 KB  conda-forge
    conda-4.6.14               |           py37_0         2.1 MB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.3 MB

The following packages will be SUPERSEDED by a higher-priority channel:

  certifi                                         pkgs/main --> conda-forge
  conda                                           pkgs/main --> conda-forge




certifi-2019.3.9     | 149 KB    |            |   0% 
certifi-2019.3.9     | 149 KB    | 8          |   8% 
certifi-2019.3.9     | 149 KB    | ########## | 100% 

conda-4.6.14         | 2.1 

In [11]:
# get the latitude and longitude values of Edmonton
address = 'Edmonton, AB'
geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude

In [15]:
# create map of Edmonton using latitude and longitude values
map_Ed = folium.Map(location=[latitude, longitude], zoom_start=10)
# add markers to map
for lat, lng, neighborhood, radd in zip(result_a['Latitude'], result_a['Longitude'], result_a['Neighbourhood Name'], result_a['Inhabitants']):
    label = '{}, {}'.format(neighborhood, radd)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=radd/500,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Ed)  
map_Ed

As we can see, there is extrimely high population density in the centre of the city. The south, the south-west and the west are also have high level of population density. The north and the north-east inhabited evenly. The east and the north-west are almost deserted.

It is time to recieve and dicover information about quantity and distribution of pharmacies in the city. The choosen cource of information is Web-service https://developer.foursquare.com. The service gives to a developer wide range of possibilities to work with API data as well as to recieve a venue information of a choosen category. In my research I will use the following categories: Pharmacy, Apotheke, Drugstore as well as citymolls (bacause all the citymolls contains a shop that sales drugs).

In [16]:
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values
# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
# tranforming json file into a pandas dataframe library

In [17]:
CLIENT_ID = '****************************' # your Foursquare ID
CLIENT_SECRET = '***********************************' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 50
category_section = '5745c2e4498e11e7bccabdbd,4bf58dd8d48988d10f951735,4bf58dd8d48988d1fd941735'
radius = 20000

In [18]:
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&categoryId={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, category_section, radius, LIMIT)

In [19]:
pharm = requests.get(url).json()

In [20]:
# assign relevant part of JSON to venues
venues = pharm['response']['venues']
# tranform venues into a dataframe
dataframe = json_normalize(venues)
dataframe.head ()

categories  hasPerk  \
0  [{'id': '4bf58dd8d48988d10f951735', 'name': 'P...    False   
1  [{'id': '4bf58dd8d48988d10f951735', 'name': 'P...    False   
2  [{'id': '4bf58dd8d48988d118951735', 'name': 'G...    False   
3  [{'id': '4bf58dd8d48988d1fd941735', 'name': 'S...    False   
4  [{'id': '4bf58dd8d48988d193941735', 'name': 'W...    False   

                         id       location.address location.cc location.city  \
0  4b0c723bf964a520743d23e3    10818 Jasper Ave NW          CA      Edmonton   
1  4b252f41f964a5208f6d24e3  1494, 8882 170 St. NW          CA      Edmonton   
2  57d021bb498e62e82ca3070f        11835 26 Ave SW          CA      Edmonton   
3  4b0586f6f964a5205b7822e3            8882 170 St          CA      Edmonton   
4  4c462592cd5ca5933defff86           8872 170 St.          CA      Edmonton   

  location.country location.crossStreet  location.distance  \
0           Canada                  NaN                622   
1           Canada   West Edmonton Mall               7982   
2           Canada         Opening soon              14295   
3           Canada     Btwn 87 & 90 Ave               7684   
4           Canada   West Edmonton Mall               7944   

                           location.formattedAddress  \
0  [10818 Jasper Ave NW, Edmonton AB T5J 2B3, Can...   
1  [1494, 8882 170 St. NW (West Edmonton Mall), E...   
2  [11835 26 Ave SW (Opening soon), Edmonton AB T...   
3  [8882 170 St (Btwn 87 & 90 Ave), Edmonton AB T...   
4  [8872 170 St. (West Edmonton Mall), Edmonton A...   

                             location.labeledLatLngs  location.lat  \
0  [{'label': 'display', 'lat': 53.54098281712579...     53.540983   
1  [{'label': 'display', 'lat': 53.52350461852903...     53.523505   
2  [{'label': 'display', 'lat': 53.40813625251965...     53.408136   
3  [{'label': 'display', 'lat': 53.52361892954877...     53.523619   
4  [{'label': 'display', 'lat': 53.52232638052472...     53.522326   

   location.lng location.neighborhood location.postalCode location.state  \
0   -113.507116                   NaN             T5J 2B3             AB   
1   -113.626965         West Edmonton             T5T 4M2             AB   
2   -113.536784                   NaN             T6W 0C8             AB   
3   -113.622429                   NaN             T5T 4J2             AB   
4   -113.626023                   NaN             T5T 4M2             AB   

                       name    referralId  
0                    Rexall  v-1556437798  
1              London Drugs  v-1556437798  
2  Real Canadian Superstore  v-1556437798  
3        West Edmonton Mall  v-1556437798  
4           World Waterpark  v-1556437798

I recieved the needed information. It contains exessive information for my research. 

In [23]:
# drop the excessive columns
dataframe = dataframe.drop (['categories', 
                            'hasPerk', 
                            'id', 
                            'location.address',
                            'location.cc',
                            'location.city',
                            'location.country',
                            'location.crossStreet',
                            'location.distance',
                            'location.formattedAddress',
                            'location.labeledLatLngs',
                            'location.postalCode',
                            'location.state',
                            'location.neighborhood',
                            'referralId'], axis = 1)
dataframe.head ()

location.lat  location.lng                      name
0     53.540983   -113.507116                    Rexall
1     53.523505   -113.626965              London Drugs
2     53.408136   -113.536784  Real Canadian Superstore
3     53.523619   -113.622429        West Edmonton Mall
4     53.522326   -113.626023           World Waterpark

For better understanding and to visualize the information show it on the map of the city.

In [25]:
# create map of Edmonton with Pharmacies
map_Edph = folium.Map(location=[latitude, longitude], zoom_start=10)
# add markers to map
for lat, lng, pharm in zip(dataframe['location.lat'], dataframe['location.lng'], dataframe['name']):
    label = '{}'.format(pharm)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='green',
        fill=True,
        fill_color='#1c701d',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Edph)  
map_Edph

It is clear that exist strong correlation beetwen distance of a neighbourhood to a nearestfrom pharmacy and the necessity in a new closer pharmacy. Subsequently, I will find the distance beetwen every neighbourhood and a nearest pharmacy.  

In [26]:
# Using geopy to calculate distance in kilometers beetwen every Neighbourhood and every Pharmacy
import geopy.distance
rr = []
for r in range(len(result_a)):
    for j in range(len(dataframe)):    
        coords_1 = (result_a.iloc[r]['Latitude'], result_a.iloc[r]['Longitude'])
        coords_2 = (dataframe.iloc[j]['location.lat'], dataframe.iloc[j]['location.lng'])
        dist = geopy.distance.distance(coords_1, coords_2).km
        rr.append([result_a.iloc[r]['Neighbourhood Name'], dist])

In [27]:
# Create DataFrame and place in it the name, the latitude and the longitude for Neighbourhoods and Pharmacies as well as distance beetwen them
df = pd.DataFrame(rr)
headers = ["Neighbourhood", "distance"]
df.columns = headers
df.head ()

Neighbourhood   distance
0     Crestwood   4.151557
1     Crestwood   4.064262
2     Crestwood  14.328593
3     Crestwood   3.776846
4     Crestwood   4.050709

I recieved the DataFrame that contains distance from every neighbourhood to every pharmacy. The next step is to find the minimum distance, convert it to metres, and cut off neighbourhoods with distance to a nearest pharmacy less than 2500 metres.

In [28]:
# Finding the minimum distance for all the Neighbourhoods and sort by distance
df_cc = df.groupby('Neighbourhood')['distance'].min().to_frame().reset_index().sort_values(by='distance')
# Convert distance to metres
df_cc['distance'] *= 1000
df_cc['distance'] = df_cc['distance'].astype(int) 
# Cut Neighbourhoods with distance to a pharmacy less than 2500 metres
df_cc = df_cc[df_cc['distance'] > 2500]

To continue the reseach I need to add more neibourhood data into the Frame. 

In [29]:
result_fn = pd.merge(left=df_cc, right=result_a, how='left', left_on='Neighbourhood', right_on='Neighbourhood Name')
result_fn = result_fn.drop (['Neighbourhood Number', 
                             'Neighbourhood Name'], axis = 1)
result_fn.head ()

Neighbourhood  distance  Area Sq Km   Latitude   Longitude  \
0            Keswick Area      2536    3.870061  53.417600 -113.635797   
1                  Fraser      2557    1.308507  53.610382 -113.374782   
2             Homesteader      2563    1.577836  53.584720 -113.406459   
3           Potter Greens      2586    1.182586  53.518290 -113.672602   
4  Maple Ridge Industrial      2594    3.745517  53.498009 -113.356365   

   Inhabitants  
0          589  
1         3489  
2         3545  
3         1490  
4           77

I recieved the DataFrame that contains all the necessary information about neighbourhoods that are away from the nearest pharmacy. 

In [31]:
# create map of Edmonton using latitude and longitude values
map_Ed = folium.Map(location=[latitude, longitude], zoom_start=10)
# add markers to map
for lat, lng, borough, neighborhood, inh in zip(result_fn['Latitude'], result_fn['Longitude'], result_fn['distance'], result_fn['Neighbourhood'], result_fn['Inhabitants']):
    label = '{}, {}'.format(neighborhood, inh)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=inh/500,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Ed)  
map_Ed

This neighbourhoods distributed in the north, the south-east and the west of Edmonton mostly. The sum of inhabatants of this neyghbourhoods is ..... From the data it can be seen that the inhabitants living close to a pharmacy is for the total number of pharmacies . It is clear, that the same proportion is used for those who live away. 

In [38]:
# Find the desired quantity of additional pharmacies
k = (len(dataframe) * result_fn['Inhabitants'].sum ())/(result_a['Inhabitants'].sum () - result_fn['Inhabitants'].sum ())
k

10.875332464393589

As we can see it shold be opened 11 (10.875 rounde to nearest) new pharmacies. 
The next problem to be solved is coordinates of the new pharmacies. The good idea in that is to use K-Means Clustering. 
I will use the initialization method of the centroids k-means++ and the number of time the k-means algorithm will be run with different centroid seeds is 10.

In [39]:
k_means = KMeans(init = "k-means++", n_clusters = 11, n_init = 10)

In [41]:
# convert the Pandas data frame to a Numpy array
X = result_fn[['Latitude', 'Longitude']].values  

In [42]:
k_means.fit(X)

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
    n_clusters=11, n_init=10, n_jobs=None, precompute_distances='auto',
    random_state=None, tol=0.0001, verbose=0)

In [43]:
# Get the coordinates of the cluster centers using KMeans
k_means_cluster_centers = k_means.cluster_centers_

In [44]:
# Create DataFrame and place in it the latitude and the longitude for new Pharmacies
df_nph = pd.DataFrame(k_means_cluster_centers)
headers = ["Latitude", "Longitude"]
df_nph.columns = headers
df_nph.head ()

Latitude   Longitude
0  53.580946 -113.387381
1  53.533322 -113.686497
2  53.416409 -113.442683
3  53.501959 -113.583066
4  53.599470 -113.547204

In [47]:
# create map of Edmonton using latitude and longitude values
map_Ednph = folium.Map(location=[latitude, longitude], zoom_start=10)
# add markers to map
for lat, lng in zip(df_nph['Latitude'], df_nph['Longitude']):
    folium.CircleMarker(
        [lat, lng],
        radius=10,
        color='blue',
        fill=True,
        fill_color='#1ed2e3',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Ednph)  
map_Ednph